File: src/lad.c
/* ID: lad.c, last updated 2020-10-09, F.Osorio */

#include "base.h"
#include "interface.h"
#include "lad.h"

/* static functions.. */
DIMS dims(int *);
void dims_free(DIMS);
double do_weight(double, double);
/* ..end declarations */

DIMS
dims(int *pdims)
{ /* dims object for linear models */
  DIMS ans;

  ans = (DIMS) Calloc(1, DIMS_struct);
  ans->n = (int) pdims[0];
  ans->p = (int) pdims[1];
  ans->N = ans->n * ans->p;
  return ans;
}

void
dims_free(DIMS this)
{ /* destructor for a dims object */
  Free(this);
}

double
do_weight(double residual, double sad)
{ /* compute weights */
  double ans, eps = R_pow(DOUBLE_EPS, .5);

  if (fabs(residual) < eps * sad)
    ans = 1.0;
  else
    ans = eps / fabs(residual);
  return ans;
}

void
lad_fitter(double *y, double *x, int *pdims, double *coef, double *scale, double *fitted,
  double *resid, double *weights, double *control, double *sad, double *logLik)
{ /* fitter for linear models under Laplace errors */
  LAD model;

  model = lad_init(y, x, pdims, coef, scale, sad, fitted, resid, weights, control);
  lad_fit(model);
  *logLik = lad_logLik(model->scale, model->dd->n);
  lad_free(model);
}

void
lad_fit(LAD model)
{
  int iter = 0;

  switch(model->method) {
    case BR:
      iter = l1fit_BR(model->y, model->x, model->dd, model->coef, model->scale, model->sad,
                      model->fitted, model->resid, model->weights, model->tolerance);
      break;
    case EM:
      iter = IRLS(model->y, model->x, model->dd, model->coef, model->scale, model->sad,
                  model->fitted, model->resid, model->weights, model->maxIter, model->tolerance);
      break;
  }

  /* save numIter */
  (model->control)[3] = (double) iter;
}

LAD
lad_init(double *y, double *x, int *pdims, double *coef, double *scale, double *sad,
  double *fitted, double *resid, double *weights, double *control)
{ /* constructor for a linear model object */
  LAD model;

  model = (LAD) Calloc(1, LAD_struct);
  model->dd = dims(pdims);
  model->y = y;
  model->x = x;
  model->coef = coef;
  model->scale = scale;
  model->sad = sad;
  model->fitted = fitted;
  model->resid = resid;
  model->weights = weights;
  model->control = control;
  model->maxIter = (int) control[0];
  model->tolerance = control[1];
  model->method = (int) control[2];
  return model;
}

void
lad_free(LAD this)
{ /* destructor for a model object */
  dims_free(this->dd);
  Free(this);
}

double
lad_objective(double *residuals, int n)
{ /* total absolute deviation */
  double ans;

  ans = sum_abs(residuals, n, 1);
  return ans;
}

double
lad_logLik(double *scale, int n)
{ /* log-likelihood function under Laplace errors */
  double ans;

  ans = (double) n * (0.5 * M_LN2 + 1.0 + log(*scale));
  return -ans;
}

int
IRLS(double *y, double *x, DIMS dd, double *coef, double *scale, double *sad, double *fitted,
  double *residuals, double *weights, int maxIter, double tolerance)
{ /* iteratively reweighted LS algorithm */
  int iter;
  double conv, SAD, newSAD, *incr, *working;

  /* initialization */
  incr    = (double *) Calloc(dd->p, double);
  working = (double *) Calloc(dd->n, double);
  SAD = lad_objective(residuals, dd->n);

  /* main loop */
  for (iter = 1; iter <= maxIter; iter++) {
    /* E-step */
    for (int i = 0; i < dd->n; i++)
      weights[i] = do_weight(residuals[i], SAD);

    /* M-step */
    IRLS_increment(y, x, dd, fitted, residuals, weights, coef, incr, working);
    newSAD = lad_objective(residuals, dd->n);
    *sad = newSAD;
    *scale = M_SQRT2 * *sad / dd->n;

    /* eval convergence */
    conv = fabs((newSAD - SAD) / (newSAD + ETA_CONV));
    if (conv < tolerance) { /* successful completion */
      Free(incr); Free(working);
      return iter;
    }
    SAD = newSAD;
  }
  Free(incr); Free(working);

  return (iter - 1);
}

void
IRLS_increment(double *y, double *x, DIMS dd, double *fitted, double *residuals,
  double *weights, double *coef, double *incr, double *working)
{ /* increment for direction search in IRLS */
  int info = 0;
  double stepsize = 1.0, wts, *z, *qraux;

  /* initialization */
  z     = (double *) Calloc(dd->n * dd->p, double);
  qraux = (double *) Calloc(dd->p, double);

  /* transformed model matrix and working residuals */
  for (int i = 0; i < dd->n; i++) {
    wts = sqrt(weights[i]);
    working[i] = wts * residuals[i];
    for (int j = 0; j < dd->p; j++)
      z[i + j * dd->n] = wts * x[i + j * dd->n];
  }

  /* QR decomposition of transformed model matrix */
  QR_decomp(z, dd->n, dd->n, dd->p, qraux, &info);
  if (info)
    error("QR_decomp in IRLS_increment gave error code %d", info);

  /* solve the transformed LS problem */
  QR_qty(z, dd->n, dd->n, dd->p, qraux, working, dd->n, dd->n, 1, &info);
  if (info)
    error("QR_qty in IRLS_increment gave error code %d", info);
  Memcpy(incr, working, dd->p);
  backsolve(z, dd->n, dd->p, incr, dd->p, 1, &info);
  if (info)
    error("DTRTRS in IRLS_increment gave error code %d", info);

  /* update coefficients */
  ax_plus_y(stepsize, incr, 1, coef, 1, dd->p);

  /* fitted values */
  for (int i = 0; i < dd->n; i++)
    fitted[i] = 0.0;
  Memcpy(fitted, coef, dd->p);
  mult_triangular_vec(z, dd->n, dd->p, "U", "N", "N", fitted, 1);
  QR_qy(z, dd->n, dd->n, dd->p, qraux, fitted, dd->n, dd->n, 1, &info);
  if (info)
    error("QR_qy in IRLS_increment gave error code %d", info);

  /* fitted values and residuals in original scale */
  for (int i = 0; i < dd->n; i++) {
    wts = sqrt(weights[i]);
    fitted[i] /= wts;
    residuals[i] = y[i] - fitted[i];
  }
  Free(z); Free(qraux);
}

int
l1fit_BR(double *y, double *x, DIMS dd, double *coef, double *scale, double *sad,
  double *fitted, double *residuals, double *weights, double tolerance)
{ /* wrapper for 'l1fit' */
  int n2 = dd->n + 2, p2 = dd->p + 2, info, iter, rank, *work;
  double minimum, *z;

  /* initialization */
  z    = (double *) Calloc(n2 * p2, double);
  work = (int *) Calloc(dd->n, int);
  copy_mat(z, n2, x, dd->n, dd->n, dd->p);

  /* call fitter */
  F77_CALL(l1fit)(z, y, &(dd->n), &(dd->p), &n2, &p2, coef, residuals, &minimum,
                  &iter, &tolerance, &rank, &info, work);
  Free(z); Free(work);

  /* save results */
  *sad   = minimum;
  *scale = M_SQRT2 * minimum / dd->n;

  /* post-processing */
  GAXPY(fitted, 1.0, x, dd->n, dd->n, dd->p, coef, 1.0);
  for (int i = 0; i < dd->n; i++) {
    /* flag for 'basic' observations */
    if (residuals[i] == 0.0)
      weights[i] = 1.0;
    else
      weights[i] = 0.0;
  }

  if (info == 0)
    error("l1fit: optimal solution is probably non-unique.");
  if (info != 1)
    error("l1fit: calculations terminated prematurely");
  if (rank != dd->p)
    error("l1fit: matrix not of full rank, apparent rank %d", rank);

  return iter;
}

File: R/lad.R
## ID: lad.R, last updated 2024-01-16, F.Osorio and T.Wolodzko

lad <-
function(formula, data, subset, na.action, method = "BR", tol = 1e-7, maxiter = 200, 
  x = FALSE, y = FALSE, contrasts = NULL)
{ ## least absolute deviations fit
  ret.x <- x
  ret.y <- y
  Call <- match.call()
  mf <- match.call(expand.dots = FALSE)
  mf$method <- mf$tol <- mf$maxiter <- mf$x <- mf$y <- mf$contrasts <- NULL
  mf$drop.unused.levels <- TRUE
  mf[[1]] <- as.name("model.frame")
  mf <- eval(mf, parent.frame())
  Terms <- attr(mf, "terms")
  y <- model.response(mf, "numeric")
  x <- model.matrix(Terms, mf, contrasts)

  # call fitter
  z <- lad.fit(x, y, method, tol, maxiter)

  # output
  z$call <- Call
  z$model <- mf
  z$na.action <- attr(mf, "na.action")
  z$contrasts <- attr(x, "contrasts")
  z$xlevels <- .getXlevels(Terms, mf)
  z$terms <- Terms
  if (ret.y)
    z$y <- y
  if (ret.x)
    z$x <- x
  class(z) <- "lad"
  z
}

lad.fit <-
function(x, y, method = "BR", tol = 1e-7, maxiter = 200)
{ ## dispatcher among various fitting functions
  if (!is.numeric(x))
    stop("model matrix must be numeric.")
  if (!is.numeric(y))
    stop("response must be numeric.")
  if (!length(x))
    method <- "null"
  contr <- attr(x, 'contrast')
  fit <- switch(method,
                BR  = lad.fit.BR(x, y, tol),
                EM  = lad.fit.EM(x, y, tol, maxiter),
                stop(paste("unimplemented method:", method)))
  fit$contrast <- contr
  fit$method <- method
  fit
}

lad.fit.BR <-
function(x, y, tol = 1e-7)
{
  if (is.matrix(y))
    stop("'lad.fit.BR' does not support multiple responses.")
  storage.mode(x) <- "double"
  storage.mode(y) <- "double"
	ny <- length(y)
	dx <- dim(x)
	n <- dx[1]
  ynames <- names(y)
	xnames <- dimnames(x)[[2]]
	if (n != ny)
		stop("Number of observations in x and y not equal.")
	p <- dx[2]
  if (n <= p)
    stop("More variables than observations.")

  # call fitter
  converged <- TRUE
  now <- proc.time()
  z <- .C("lad_BR",
          y = y,
          x = x,
          n = as.integer(n),
          p = as.integer(p),
          coef = double(p),
          scale = as.double(0),
          sad = as.double(0),
          fitted = double(n),
          residuals = double(n),
          logLik = as.double(0),
          tol = as.double(tol),
          rank = as.integer(0),
          iter = as.integer(0),
          info = as.integer(0))
  speed <- proc.time() - now
  msg <- character(0)
  if (z$info == 0)
    msg <- "Non-unique solution possible."
  else if(z$info != 1) {
    msg <- "Premature termination."
    converged <- FALSE
  }
  if (z$rank != p)
    msg <- c(msg, paste("Matrix not of full rank, apparent rank", z$rank))

  # post-processing
  basic <- (1:n)[z$residuals == 0.0]
  R <- chol(crossprod(x))

  # creating the output object
  out <- list(coefficients = z$coef, scale = z$scale, residuals = z$residuals,
              fitted.values = z$fitted, SAD = z$sad, logLik = z$logLik, basic = basic,
              dims = dx, R = R, iterations = z$iter, speed = speed, converged = converged)
  if (length(msg))
    out$message <- msg
  names(out$coefficients) <- xnames
  names(out$residuals) <- ynames
  names(out$fitted.values) <- ynames
  dimnames(out$R) <- list(xnames, xnames)
  class(out) <- "lad"
  out
}

lad.fit.EM <-
function(x, y, tol = 1e-7, maxiter = 200)
{
  if (is.matrix(y))
    stop("'lad.fit.EM' does not support multiple responses.")
  storage.mode(x) <- "double"
  storage.mode(y) <- "double"
	ny <- length(y)
	dx <- dim(x)
	n <- dx[1]
  ynames <- names(y)
	xnames <- dimnames(x)[[2]]
	if (n != ny)
		stop("Number of observations in x and y not equal.")
	p <- dx[2]
  if (n <= p)
    stop("More variables than observations.")

  # set control values
  control <- c(maxiter, tol, 0)

  # initial estimates
  fit <- lsfit(x, y, intercept = FALSE)
  res <- fit$residuals
  cf <- fit$coefficients
  R <- qr.R(fit$qr)

  # call fitter
  converged <- FALSE
  now <- proc.time()
  z <- .C("lad_EM",
          y = y,
          x = x,
          n = as.integer(n),
          p = as.integer(p),
          coefficients = as.double(cf),
          scale = as.double(0),
          fitted = double(n),
          residuals = as.double(res),
          weights = as.double(rep(1, n)),
          sad = as.double(0),
          logLik = as.double(0),
          tol = as.double(tol),
          iter = as.integer(maxiter))
  speed <- proc.time() - now

  # post-processing
  if (z$iter < maxiter)
    converged <- TRUE
  eps <- .Machine$double.eps^.35
  dev <- z$residuals
  test <- abs(dev) < eps * z$sad
  weights <- ifelse(test, 1.0, eps / abs(dev))
  basic <- (1:n)[test]
  msg <- character(0)
  if (length(basic) == 0) {
    basic <- NULL
    msg <- "method was unable to determine basic observations"
  }

  # creating the output object
  out <- list(coefficients = z$coef, scale = z$scale, residuals = z$residuals,
              fitted.values = z$fitted, SAD = z$sad, logLik = z$logLik,
              weights = weights, basic = basic, dims = dx, R = R, iterations = z$iter,
              speed = speed, converged = converged)
  if (length(msg))
    out$message <- msg
  names(out$coefficients) <- xnames
  names(out$residuals) <- ynames
  names(out$fitted.values) <- ynames
  names(out$weights) <- ynames
  dimnames(out$R) <- list(xnames, xnames)
  class(out) <- "lad"
  out
}

print.lad <-
function(x, digits = 4, ...)
{
  cat("Call:\n")
  dput(x$call)
  if (x$converged)
    cat("Converged in", x$iterations, "iterations\n")
  else {
    if (x$method == "EM")
      cat("Maximum number of iterations exceeded\n")
  }
  cat("\nCoefficients:\n")
  print(format(round(x$coef, digits = digits)), quote = F, ...)
  nobs <- x$dims[1]
  rdf <- nobs - x$dims[2]
  cat("\nDegrees of freedom:", nobs, "total;", rdf, "residual")
  cat("\nScale estimate:", format(x$scale), "\n")
  if (!is.null(x$message))
    cat(paste("\nNOTE:", x$message, "\n"))
  invisible(x)
}

residuals.lad <-
function(object, type = c("working", "response", "quantile"), ...)
{
  type <- match.arg(type)
  r <- object$residuals
  res <- switch(type,
                working = r, response = r,
                quantile = {
                  y <- model.response(object$model)
                  mu <- fitted(object)
                  dispersion <- object$scale
                  logp <- plaplace(y, location = mu, scale = dispersion, log.p = TRUE)
                  qnorm(logp, log.p = TRUE)
                })
  res
}

fitted.lad <- function(object, ...)
  object$fitted.values

## sqrt(2) 'factor' has not been included
deviance.lad <- function(object, ...)
  sum(abs(residuals(object)), na.rm = TRUE)

## log-likelihood for lad objects
logLik.lad <- function(object, ...)
{
  val <- object$logLik
  N <- object$dims[1]
  p <- object$dims[2]
  attr(val, "nall") <- N
  attr(val, "nobs") <- N # basic observations must be deleted?
  attr(val, "df") <- p + 1
  class(val) <- "logLik"
  val
}

predict.lad <- function(object, newdata, na.action = na.pass, ...)
{
  if (missing(newdata) || is.null(newdata))
    return(fitted(object))
  tt <- terms(object)
  Terms <- delete.response(tt)
  m <- model.frame(Terms, newdata, na.action = na.action, xlev = object$xlevels)
  if (!is.null(cl <- attr(Terms, "dataClasses")))
    .checkMFClasses(cl, m)
  X <- model.matrix(Terms, m, contrasts.arg = object$contrasts)
  drop(X %*% coef(object))
}

plot.lad <-
function(x, which = 1:2,
         caption = c("Residuals vs Fitted", "Laplace Q-Q plot"),
         panel = points,
         sub.caption = deparse(x$call), main = "",
         ask = prod(par("mfcol")) < length(which) && dev.interactive(),
         ...,
         id.n = 3, labels.id = names(residuals(x)), cex.id = 0.75)
{
  if (!inherits(x, "lad"))
    stop("Use only with 'lad' objects")
  if (!is.numeric(which) || any(which < 1) || any(which > 2))
    stop("`which' must be in 1:2")
  show <- rep(FALSE, 2)
  show[which] <- TRUE
  rs <- residuals(x, type = "quantile") # a kind of standardized residual
  yhat <- fitted(x)
  rs.lab <- "Standardized residuals"
  yh.lab <- "Fitted values"
  n <- length(rs)
  if (is.null(id.n))
    id.n <- 0
  else {
    id.n <- as.integer(id.n)
    if (id.n < 0 || id.n > n)
      stop("`id.n' must be in {1,..,",n,"}")
  }
  if (id.n > 0) { # label the largest residuals
    if (is.null(labels.id))
      labels.id <- paste(1:n)
    iid <- 1:id.n
    show.rs <- sort.list(abs(rs), decreasing = TRUE)[iid]
    text.id <- function(x, y, ind, adj.x = FALSE)
      text(x - if(adj.x) strwidth(" ")*cex.id else 0, y, labels.id[ind],
           cex = cex.id, xpd = TRUE, adj = if(adj.x) 1)
  }
  one.fig <- prod(par("mfcol")) == 1
  if (ask) {
    op <- par(ask = TRUE)
    on.exit(par(op))
  }
  #---------- Do the individual plots : ----------
  if (show[1L]) {
    ylim <- range(rs, na.rm = TRUE)
    if (id.n > 0)
      ylim <- ylim + c(-1,1)* 0.08 * diff(ylim)
    plot(yhat, rs, xlab = yh.lab, ylab = rs.lab, main = main, ylim = ylim, type = "n", ...)
    panel(yhat, rs, ...)
    if (one.fig)
      title(sub = sub.caption, ...)
    mtext(caption[1], 3, 0.25)
    if (id.n > 0) {
      y.id <- rs[show.rs]
      y.id[y.id < 0] <- y.id[y.id < 0] - strheight(" ")/3
      text.id(yhat[show.rs], y.id, show.rs, adj.x = TRUE)
    }
    abline(h = 0, lty = 3, col = "gray")
  }
  if (show[2L]) { # Laplace
    ylim <- range(rs, na.rm=TRUE)
    ylim[2] <- ylim[2] + diff(ylim) * 0.075
    qq <- qqnorm(rs, main = main, ylab = rs.lab, ylim = ylim, ...)
    abline(c(0,1), lty = 3, col = "gray")
    if (one.fig)
      title(sub = sub.caption, ...)
    mtext(caption[2], 3, 0.25)
    if (id.n > 0)
      text.id(qq$x[show.rs], qq$y[show.rs], show.rs, adj.x = TRUE)
  }
  if (!one.fig && par("oma")[3] >= 1)
  mtext(sub.caption, outer = TRUE, cex = 1.25)
  invisible()
}

simulate.lad <- function(object, nsim = 1, seed = NULL, ...)
{
  if (!exists(".Random.seed", envir = .GlobalEnv, inherits = FALSE))
    runif(1)  # initialize the RNG if necessary
  if (is.null(seed))
    RNGstate <- get(".Random.seed", envir = .GlobalEnv)
  else {
    R.seed <- get(".Random.seed", envir = .GlobalEnv)
    set.seed(seed)
    RNGstate <- structure(seed, kind = as.list(RNGkind()))
    on.exit(assign(".Random.seed", R.seed, envir = .GlobalEnv))
  }
  ftd <- fitted(object)
  nm <- names(ftd)
  n <- length(ftd)
  ntot <- n * nsim
  ans <- as.data.frame(ftd +
                       matrix(rlaplace(ntot, scale = object$scale), nrow = n))
  names(ans) <- paste("sim", seq_len(nsim), sep = "_")
  if (!is.null(nm))
    row.names(ans) <- nm
  attr(ans, "seed") <- RNGstate
  ans
}

confint.lad <- function(object, parm, level = 0.95, ...)
{
  cf <- coef(object)
  pnames <- names(cf)
  if (missing(parm)) parm <- seq(along = pnames)
  else if(is.character(parm)) parm <- match(parm, pnames, nomatch = 0)
  a <- (1 - level) / 2
  a <- c(a, 1-a)
  pct <- paste(round(100 * a, 1), "%")
  ci <- array(NA, dim = c(length(parm), 2),
              dimnames = list(pnames[parm], pct))
  se <- sqrt(diag(vcov(object)))[parm]
  qz <- qnorm(a)
  ci[] <- cf[parm] + se %o% qz
  ci
}

vcov.lad <- function(object, ...)
{
  scale <- object$scale
  2.0 * scale^2 * chol2inv(object$R)
}

summary.lad <-
function (object, ...)
{
  z <- object
  p <- z$dims[2]
  cov.unscaled <- chol2inv(z$R)
  se <- z$scale * sqrt(2.0 * diag(cov.unscaled))
  est <- z$coefficients
  zval <- est / se
  ans <- z[c("call", "terms")]
  ans$dims <- z$dims
  ans$logLik <- z$logLik
  ans$scale <- z$scale
  ans$residuals <- z$residuals
  ans$coefficients <- cbind(est, se, zval, 2 * pnorm(abs(zval), lower.tail = FALSE))
  dimnames(ans$coefficients) <- list(names(z$coefficients),
        c("Estimate", "Std.Error", "Z value", "p-value"))
  ans$cov.unscaled <- cov.unscaled
  class(ans) <- "summary.lad"
  ans
}

print.summary.lad <-
function(x, digits = 4, ...)
{
  cat("Call:\n")
  dput(x$call)
  resid <- x$residuals
  nobs <- x$dims[1]
  p <- x$dims[2]
  rdf <- nobs - p
  if (rdf > 5) {
    cat("\nResiduals:\n")
		rq <- quantile(resid)
		names(rq) <- c("Min", "1Q", "Median", "3Q", "Max")
		print(rq, digits = digits, ...)
	}
	else if(rdf > 0) {
	 cat("\nResiduals:\n")
	 print(resid, digits = digits, ...)
  }
  cat("\nCoefficients:\n")
  print(format(round(x$coef, digits = digits)), quote = F, ...)
  cat("\nDegrees of freedom:", nobs, "total;", rdf, "residual")
  cat("\nScale estimate:", format(x$scale))
  cat("\nLog-likelihood:", format(x$logLik), "on", p + 1, "degrees of freedom\n")
  invisible(x)
}
